sdk/python/src/timecraft/app.py

```diff
@@ -0,0 +1,212 @@
+    """
+    _func: Callable[..., Any]
+    _app: "App"
+    _name: str
+
+    def __init__(self, app: "App", name: str, f: Callable[..., Any]) -> None:
+        self._func = f
+        self._app = app
+        self._name = name
+
+    def __call__(self, *args):
+        return self._func(*args)
+
+    def call(self, *args, **kwargs) -> Promise:
+        """
+        Calls the function with the given arguments and returns a Promise.
+        """
+        return self._app._spawn_task(self._name, *args, **kwargs)
+
+
+class App:
+    """
+    App is the main class of the Timecraft Python engine. An App will be used
+    to register a set of functions to be scheduled by the Ring scheduler.
+    """
+    _name: Optional[str]
+    _client: Client
+
+    _start: Entrypoint
+    _functions: Dict[str, Function]
+
+    _web: aiohttp.web.Application
+
+    def __init__(self) -> None:
+        self._tasks = {}
+        self._client = Client()
+        self._functions = {}
+
+        # TODO: move somewhere else
+        self._web = aiohttp.web.Application()
+        self._web.add_routes([aiohttp.web.post("/", self._handle)])
+
+    async def _serve(self):
+        # TODO: handle failures
+        await aiohttp.web._run_app(app=self._web, host="0.0.0.0", port=3000)
+
+    # TODO: ensure the HTTP server can be gracefully shutted down once
+    # the main function is closed.
+    async def run(self, args):
+        """
+        Run the app. If no arguments are passed, the entrypoint function will
+        be executed. Otherwise, the app will start an HTTP server waiting for
+        task to be submitted.
+        """
+        if len(args) == 0:
+            await self._start()
+            return
+
+        await self._serve()
+
+    def function(self, name: Optional[str] = None) -> Callable[..., Function]:
+        """
+        Function decorator used to register a function to the app.
+        """
+        def decorator(func: Callable[..., Any]) -> Function:
+            if name:
+                fid = name
+            else:
+                m = hashlib.sha256()
+                m.update(cloudpickle.dumps(func))
+                fid = m.hexdigest()
+
+            f = Function(self, fid, func)
+            self._functions[fid] = f
+            self._web.add_routes([aiohttp.web.post(f"/{fid}", self._handle)])
+            return f
+        return decorator
+
+    def start(self) -> Callable[..., Any]:
+        """
+        Decorator used to register the entrypoint function of the app.
+        Raise an exception is called multiple times.
+        """
+        if self._start is not None:
+            raise Exception("start function already defined")
+
+        def decorator(func: Callable[..., Any]) -> Callable[..., Any]:
+
+            self._start = Entrypoint(func, self)
+            return func
+        return decorator
+
+    def _spawn_task(self, name: str, *args, **kwargs) -> Promise:
+        workerfile = os.path.abspath(__file__)
```

Comment:
This is quite fragile. Can we leverage our OCI layer to avoid passing random .py files?